# Homework 8: Logistic Regression
Data Science and Applied Machine Learning \\
Dr. Kerby \\
Student Name: Ziming Wang

### Question 1
Take your dataset from the mid-project (the final dataset with the most features, including your engineered features). Train a logistic regression ML model on your data to predict whether the home or visitor team will win.   


In [39]:
import numpy as np
import pandas as pd
import matplotlib
import matplotlib.pyplot as plt
%matplotlib inline
import math
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.linear_model import LogisticRegression

In [40]:
# Import the csv file
sports = pd.read_csv('DataSet_With.csv')
sports

Visitor Team             Home Team  ... VisitorWins2019  HomeWin
0     New Orleans Pelicans       Toronto Raptors  ...              33        1
1       Los Angeles Lakers  Los Angeles Clippers  ...              37        1
2            Chicago Bulls     Charlotte Hornets  ...              22        1
3          Detroit Pistons        Indiana Pacers  ...              41        0
4      Cleveland Cavaliers         Orlando Magic  ...              19        1
...                    ...                   ...  ...             ...      ...
1138            Miami Heat    Los Angeles Lakers  ...              39        1
1139    Los Angeles Lakers            Miami Heat  ...              37        1
1140    Los Angeles Lakers            Miami Heat  ...              37        0
1141            Miami Heat    Los Angeles Lakers  ...              39        0
1142    Los Angeles Lakers            Miami Heat  ...              37        0

[1143 rows x 8 columns]

In [41]:
# Explore dataset
sports.value_counts()

Visitor Team           Home Team         Overtime_OT  HomeNumWins  VisitorNumWins  HomeWins2019  VisitorWins2019  HomeWin
Boston Celtics         Toronto Raptors   0            60           58              58            49               0          6
Los Angeles Clippers   Dallas Mavericks  0            45           56              33            48               0          5
Oklahoma City Thunder  Houston Rockets   0            49           47              53            49               1          5
Milwaukee Bucks        Orlando Magic     0            34           61              42            60               0          4
Indiana Pacers         Miami Heat        0            58           45              39            48               1          4
                                                                                                                            ..
New York Knicks        Detroit Pistons   0            20           21              41            17               1 

In [42]:
# Quzntize the dataset
sports = pd.get_dummies(sports)
sports

HomeNumWins  VisitorNumWins  ...  Overtime_OT_2OT  Overtime_OT_OT
0              60              30  ...                0               1
1              56              68  ...                0               0
2              23              22  ...                0               0
3              45              20  ...                0               0
4              34              19  ...                0               0
...           ...             ...  ...              ...             ...
1138           68              58  ...                0               0
1139           58              68  ...                0               0
1140           58              68  ...                0               0
1141           68              58  ...                0               0
1142           58              68  ...                0               0

[1143 rows x 68 columns]

In [52]:
# Full feature set logistic regression
X = sports.drop('HomeWin', axis = 1)
y = sports['HomeWin']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 1/2)
X_train

HomeNumWins  VisitorNumWins  ...  Overtime_OT_2OT  Overtime_OT_OT
1023           58              34  ...                0               0
860            37              58  ...                0               0
989            34              45  ...                0               0
257            60              21  ...                0               0
53             68              34  ...                0               0
...           ...             ...  ...              ...             ...
661            49              55  ...                0               0
355            19              49  ...                0               0
525            20              45  ...                0               0
430            35              20  ...                0               0
957            47              60  ...                0               0

[571 rows x 67 columns]

In [44]:
full_log_reg = LogisticRegression()
full_log_reg.fit(X_train,y_train)

/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [45]:
print('Training accuracy: ', full_log_reg.score(X_train,y_train))
print('Testing accuracy: ', full_log_reg.score(X_test,y_test))

Training accuracy:  0.6970227670753065
Testing accuracy:  0.6590909090909091


In [46]:
pd.Series(data=full_log_reg.coef_[0], index=X_train.columns).sort_values(ascending=False)

Home Team_Philadelphia 76ers         1.237116
Home Team_San Antonio Spurs          0.643004
Visitor Team_San Antonio Spurs       0.636672
Home Team_Miami Heat                 0.596972
Home Team_Chicago Bulls              0.564168
                                       ...   
Visitor Team_Charlotte Hornets      -0.594847
Home Team_Charlotte Hornets         -0.723346
Visitor Team_New Orleans Pelicans   -0.733104
Home Team_Denver Nuggets            -0.809624
Visitor Team_Toronto Raptors        -1.089766
Length: 67, dtype: float64

### Question 2
What is the training and testing accuracy of your logistic regression prediction model? \\
How does this compare to the linear regression and random forests (or xgboost) model from the mid-project? \\
What, according to the logistic regression model, are the most important features? 

1. This model shows about 0.70 accuracy on the training data and 0.648 accuracy on the testing data.  
2. From the results, Logistic regression ML model's prediction accuracy is similar to that of the random forest model. It is better than the linear regression model.  
3. From the above code, it is difficult to determine which feature is important. The size of these coefficients does not tell me the relative importance of the model. Because it may be a relative feature scale.  

### Question 3

Normalize your feature set with StandardScaler and repeat questions 1 and 2.

In [47]:
# Normalize my feature set with StandardScaler
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_train

array([[ 0.48721243,  0.61211077,  0.75114232, ...,  0.26701041,
        -0.07267524, -0.25553573],
       [ 0.62073693,  0.48137755,  1.01187318, ...,  0.26701041,
        -0.07267524, -0.25553573],
       [-0.58098358,  1.85407631, -0.20487083, ...,  0.26701041,
        -0.07267524, -0.25553573],
       ...,
       [ 0.48721243, -0.30302173,  0.75114232, ...,  0.26701041,
        -0.07267524, -0.25553573],
       [ 0.62073693,  0.48137755,  1.01187318, ...,  0.26701041,
        -0.07267524, -0.25553573],
       [ 0.35368793, -0.36838834,  0.57732175, ...,  0.26701041,
        -0.07267524, -0.25553573]])

In [48]:
scaled_log_reg = LogisticRegression()
scaled_log_reg.fit(X_train, y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [49]:
# Have to scale our X_test first before we use it
X_test = scaler.transform(X_test)
print('Training accuracy: ', scaled_log_reg.score(X_train,y_train))
print('Testing accuracy: ', scaled_log_reg.score(X_test,y_test))

Training accuracy:  0.712784588441331
Testing accuracy:  0.6520979020979021


In [50]:
pd.Series(data=scaled_log_reg.coef_[0], index=X.columns).sort_values(ascending=False).round(3)

Home Team_Philadelphia 76ers       0.346
HomeNumWins                        0.280
Home Team_Miami Heat               0.238
Visitor Team_San Antonio Spurs     0.191
Visitor Team_Detroit Pistons       0.177
                                   ...  
Visitor Team_Los Angeles Lakers   -0.200
Home Team_Detroit Pistons         -0.201
Home Team_Charlotte Hornets       -0.208
VisitorNumWins                    -0.385
Visitor Team_Toronto Raptors      -0.393
Length: 67, dtype: float64

1. This model shows about 0.71 accuracy on the training data and 0.65 accuracy on the testing data.  
2. From the results, Logistic regression ML model's prediction accuracy is similar to that of the random forest model. It is better than the linear regression model.  
3. By using StandardScaler to normalize my feature set, I found the important features from high to low: HomeNumWins, Home Team_Philadelphia 76ers, Visitor Team_Philadelphia 76ers, Home Team_Miami Heat, Visitor Team_Golden State Warriors. These five characteristics are the most important for home team wins.  

  I also found five characteristics which are the most important for visitor team wins. From high to low: Visitor Team_Toronto Raptors, VisitorNumWins, Home Team_Charlotte Hornets, Home Team_Detroit Pistons, Visitor Team_Los Angeles Lakers. They are also the 5 features that are the largest obstacle to home team wins.  